### PR TITLE
(py)Brush v1.0: islands!

### PR DESCRIPTION
**Pull request by gAldeia**
# (py)Brush v1.0: islands!
-----

## What's new

Essentially, this PR implements all the evolutionary steps in C++, conveniently wrapped into a scikit estimator. C++ implementation uses task flow to manage several islands into different threads.

There are several new classes, and I have restructured the source code.

I named the C++ implementation _Brush_, while the Python library is called _Pybrush_.

## How is designed

The main entry point is what I call the Brush `Engine`, which will do all the work to get the job done. The Engine is configured through a struct called `Parameters`, which contains all hyper-parameters for the EA. The `Dataset` class handles every operation regarding the data (splitting train and test partitions, inferring data types, etc.). You can run the Engine using a pre-constructed Dataset instance, or you can conveniently call `fit(X, y)`, and it will try to create the dataset for you using some configurations you may specify in the Parameters class.

## Prototyping with Brush

There is still compatibility with DEAP to prototype different evolutionary algorithms. In fact, this compatibility is extended by creating binders to all classes from Brush: `Evaluator`, `Selector`, `Population`, `Individual`, etc. Brush implements hashes for the individual and fitness classes in C++, making it possible to use the DEAP toolbox. The old NSGA2 implemented using the DEAP API is still there, and in the future, I hope I can create a notebook explaining how to prototype with Brush.

## Final remarks

While there is much work to do, I think it is already time to move all these significant changes to the master branch. We now have a fully functional evolutionary algorithm implementation in C++ with taskflow to handle parallelism. There is also a convenient wrapper for this implementation, and we are still compatible with DEAP. There are many TODOs written all over the place, and I intend to work on these in the next weeks.

### CODE REVIEW

**Comment by gAldeia** [comment]
PR

File: src/eval/metrics.cpp [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+{
+        
+    /* std::cout << "loss: " << loss.transpose() << "\n"; */
+    loss = log_loss(y,predict_proba,class_weights);
+    return loss.mean();
+}
+
+
+// multinomial log loss
+VectorXf multi_log_loss(const VectorXf& y, const ArrayXXf& predict_proba,
```
Comment (lacava) [inline src/eval/metrics.cpp:63]: ?!

File: src/ind/fitness.h [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    std::size_t operator()(const std::vector<float>& v) const {
+        std::size_t seed = v.size();
+        for (const auto& elem : v) {
+            seed ^= std::hash<float>{}(elem) +  0x9e3779b9 + (seed <<  6) + (seed >>  2);
+        }
+        return seed;
+    }
+};
+
+// TODO: separate declaration from implementation (for all classes. have a folder with headers and other with srcs, just like operon)
```
Comment (lacava) [inline src/ind/fitness.h:22]: don't do that please (it's annoying)

Comment (lacava) [reply src/ind/fitness.h:22]: my opinion of course

File: src/ind/individual.h [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+    void set_fitness(Fitness &f) { fitness=f; };
+    Fitness& get_fitness() { return fitness; };
+
+    // TODO: USE setters and getters intead of accessing it directly
+    // template<ProgramType T>
+    // void Individual<T>::set_objectives(const vector<string>& objectives)
+
+    // Static map for weights associated with strings
+    inline static std::map<std::string, float> weightsMap = []() {
```
Comment (lacava) [inline src/ind/individual.h:94]: should this be in `Individual`?

**Review by lacava — COMMENTED** [review]
mostly minor comments. thanks a lot guilherme! i will merge once i hear back about the questions.

**Comment by lacava** [comment]
@gAldeia checking back on this merge

File: README.md [hunk 4]
```diff
@@ -93,6 +93,8 @@ That means it should be compatible with sklearn pipelines, wrappers, and so fort
 
 In addition, Brush provides functionality that allows you to feed in more complicated data types than just matrices of floating point values. 
 
+<!-- TODO: update these examples (at least check they still work) -->
```
[the remark this answers is not in the document]
Comment (gAldeia) [reply README.md:96]: I am going to update it so the examples work.
Just an observation: I renamed it to pybrush instead of brush, because Joe said (a few months ago) that there was already a package named brush!

[the remark this answers is not in the document]
Comment (lacava) [reply README.md:96]: the python package is already called pybrush https://github.com/cavalab/brush/blob/f54104ff12fb9079c90fabe16defad13f751b8c9/setup.py#L109

File: src/eval/metrics.cpp (hunk 1 above)
Comment (gAldeia) [reply src/eval/metrics.cpp:63]: This is a placeholder for multi classification, as softmax is not working properly. I will eventually work on this

File: src/engine.h
[the remark this answers is not in the document]
Comment (gAldeia) [reply src/engine.h:1]: Yes, it is the brushEstimator backend in c++! 

I will add docscrings before merge.

@lacava Is there any instructions on how to build the docs? you can vaguely explain as a comment here and I can try to make it work (just so I can check how they are looking). I can also write a proper guide on how to build the docs on README

File: pybrush/BrushEstimator.py
[the remark this answers is not in the document]
Comment (gAldeia) [reply pybrush/BrushEstimator.py:1]: This is actually a good idea. It will help me develop parallel classes without worrying about changing one but not the other.
I am going to create a Python base class (something like `EstimatorInterface`) and make them inherit from it. Does that sound good?

File: pybrush/__init__.py
[the remark this answers is not in the document]
Comment (gAldeia) [reply pybrush/__init__.py:1]: They compute and assign the fitness values. I was not sure if I was exposing the C++ implementation too much when doing these. I am going to leave a comment here to figure this out later, ok?

[the remark this answers is not in the document]
Comment (lacava) [reply pybrush/__init__.py:1]: yes

File: src/engine.h
[the remark this answers is not in the document]
Comment (lacava) [reply src/engine.h:1]: https://www.sphinx-doc.org/en/master/man/sphinx-build.html

**Comment by gAldeia** [comment]
@lacava 

I finished working on your comments. There is documentation for many of the new things I implemented in this PR, mostly done with the help of GitHub Copilot (I was exploring its capabilities, and it turns out it can write documentation—sometimes). I also cleaned up a lot of TODOs that I left in the code. These TODOs also help make it easier to implement the MABs, which I'll be working on now.

Some new things:

- Saving/loading population;
- Archive and predict with archive;
- New example notebooks https://github.com/cavalab/brush/tree/island_GA/docs/guide;
- Average precision score metric for binary classification.

These additions were staged locally in my machine, and while I was cleaning some TODOs, I decided to include them in the PR as well.
